integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/InVMKafkaCluster.java

```diff
@@ -0,0 +1,194 @@
+                        throw new IllegalStateException();
+                    }
+                    var port = controllerPorts.get(brokerNum);
+                    return EndpointPair.builder().bind(new Endpoint("0.0.0.0", port)).connect(new Endpoint("localhost", port)).build();
+                }
+            };
+
+            servers = clusterConfig.getBrokerConfigs(kafkaEndpointsSupplier, zookeeperEndpointSupplier).map(this::buildKafkaServer).collect(Collectors.toList());
+
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @NotNull
+    private Server buildKafkaServer(ConfigHolder c) {
+        bootstraps.add(c.getEndpoint());
+        KafkaConfig config = buildBrokerConfig(c, tempDirectory);
+        Option<String> threadNamePrefix = Option.apply(null);
+
+        boolean kraftMode = clusterConfig.isKraftMode();
+        if (kraftMode) {
+            var directories = StorageTool.configToLogDirectories(config);
+            var clusterId = c.getKafkaKraftClusterId();
+            var metaProperties = StorageTool.buildMetadataProperties(clusterId, config);
+            StorageTool.formatCommand(System.out, directories, metaProperties, true);
+            return new KafkaRaftServer(config, Time.SYSTEM, threadNamePrefix);
+        }
+        else {
+            return new KafkaServer(config, Time.SYSTEM, threadNamePrefix, false);
+
+        }
+    }
+
+    @NotNull
+    private KafkaConfig buildBrokerConfig(ConfigHolder c, Path tempDirectory) {
+        Properties properties = new Properties();
+        properties.putAll(c.getProperties());
+        Path logsDir = tempDirectory.resolve(String.format("broker-%d", c.getBrokerNum()));
+        properties.setProperty(KafkaConfig.LogDirProp(), logsDir.toAbsolutePath().toString());
+
+        return new KafkaConfig(properties);
+    }
+
+    @Override
+    public void start() {
+        if (zooFactory != null) {
+            try {
+                zooFactory.startup(zooServer);
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                zooServer.shutdown(true);
+                return;
+            }
+        }
+
+        servers.stream().parallel().forEach(Server::startup);
+
+    }
+
+    @Override
+    public String getBootstrapServers() {
+        return String.join(",", bootstraps);
+    }
+
+    @Override
+    public Map<String, Object> getConnectConfigForCluster() {
+        return clusterConfig.getConnectConfigForCluster(getBootstrapServers());
+    }
+
+    @Override
+    public void close() throws Exception {
+        try {
+            servers.stream().parallel().forEach(Server::shutdown);
+            bootstraps.clear();
+
+            if (zooServer != null) {
+                zooServer.shutdown(true);
+            }
```

Comment:
Server#shutdown will throw exception if any. We should catch throwable, otherwise, `zooServer` will have resource leakage.

Reply:
I've used a try/finally